### DISCUSSION TITLE
[V4] Setting Google provider access type to offline for refresh token causes JWT session error

### DISCUSSION THREAD

**Discussion by agusterodin** [nextauthjs/next-auth#3016] [Help] [answered]
### Question 💬

I am attempting to use the V4 beta with Google as a provider and with refresh token rotation. I had to improvise since it appears that https://next-auth.js.org/tutorials/refresh-token-rotation is outdated. Logging in with Google works fine but I get this error the next time I reach out to next auth (through use of getSession): 

```
[next-auth][error][JWT_SESSION_ERROR] 
https://next-auth.js.org/errors#jwt_session_error signature verification failed {
  message: 'signature verification failed',
  stack: 'JWSVerificationFailed: signature verification failed\n' +
    '    at jwsVerify (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/jose/lib/jws/verify.js:159:13)\n' +
    '    at Object.module.exports [as verify] (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/jose/lib/jwt/verify.js:235:5)\n' +
    '    at Object.decode (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next-auth/jwt/index.js:100:28)\n' +
    '    at Object.session (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next-auth/server/routes/session.js:39:38)\n' +
    '    at NextAuthHandler (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next-auth/server/index.js:145:29)\n' +
    '    at runMicrotasks (<anonymous>)\n' +
    '    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n' +
    '    at async /Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next-auth/server/index.js:251:32\n' +
    '    at async Object.apiResolver (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next/dist/server/api-utils.js:101:9)\n' +
    '    at async DevServer.handleApiRequest (/Users/Jeff/Documents/Pixelboard/pixelboard/ui/node_modules/next/dist/server/next-server.js:770:9)',
  name: 'JWSVerificationFailed'
}
```

### How to reproduce ☕️

I have tried this

```
export default NextAuth({
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
      authorization: {
        params: { scope: 'openid email profile', access_type: 'offline' }
      }
    })
  ]
})

```
and this

```
export default NextAuth({
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID as string,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
      authorization:
        'https://accounts.google.com/o/oauth2/v2/auth?' +
        new URLSearchParams({
          prompt: 'consent',
          access_type: 'offline',
          response_type: 'code'
        })
    })
  ]})
```

And both yield the error.

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ConciergeMento** [comment, marked as answer]
@agusterodin i am doing something similar as well, if you find a work around can you post it here as well?

**Reply by agusterodin** [reply]
Unfortunately no. I haven't had too much of a chance to play with it though.

@balazsorban44, do you have any hints or first intuitions to help us get started? Does Google's well-known not support offline tokens? Is this an issue with other providers to your knowledge?

**Reply by balazsorban44** [reply]
if I recall correctly, Google uses the same refresh token again and again https://developers.google.com/identity/protocols/oauth2#5.-refresh-the-access-token,-if-necessary.


it's less secure than rotating the refresh token every time you refresh the access token as well, but I don't think you get a choice with Google

**Reply by agusterodin** [reply]
Here is a minimal reproduction: https://github.com/agusterodin/next-auth-google-offline-token-bug-reproduction

I am simply trying to get the example from the docs working with v4 (https://github.com/lawrencecchen/next-auth-refresh-tokens) from here https://next-auth.js.org/tutorials/refresh-token-rotation. 

I commented out the callback stuff to just boil it down to the core issue. Adding callbacks back in will be the next logical step.

This seems like a major regression for v4 and I would imagine a lot of people would want to be able to refresh tokens in their web app (especially with a provider as popular as Google).

I would gladly create a pull request to update the docs as well as provide a fully working example repository updated to use v4.

**Reply by agusterodin** [reply]
The error appears when I either manually call getSession at any point in time or if I stop and restart the UI process

Also the error is slightly different this time around

```
[next-auth][error][JWT_SESSION_ERROR] 
https://next-auth.js.org/errors#jwt_session_error decryption operation failed {
  message: 'decryption operation failed',
  stack: 'JWEDecryptionFailed: decryption operation failed\n' +
    '    at gcmDecrypt (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/jose/dist/node/cjs/runtime/decrypt.js:67:15)\n' +
    '    at decrypt (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/jose/dist/node/cjs/runtime/decrypt.js:92:20)\n' +
    '    at flattenedDecrypt (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/jose/dist/node/cjs/jwe/flattened/decrypt.js:119:52)\n' +
    '    at async compactDecrypt (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/jose/dist/node/cjs/jwe/compact/decrypt.js:18:23)\n' +
    '    at async jwtDecrypt (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/jose/dist/node/cjs/jwt/decrypt.js:8:23)\n' +
    '    at async Object.decode (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/next-auth/jwt/index.js:60:7)\n' +
    '    at async Object.session (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/next-auth/core/routes/session.js:38:28)\n' +
    '    at async NextAuthHandler (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/next-auth/core/index.js:65:16)\n' +
    '    at async NextAuthNextHandler (/Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/next-auth/next/index.js:40:7)\n' +
    '    at async /Users/Jeff/Downloads/next-auth-refresh-tokens-main/node_modules/next-auth/next/index.js:80:32',
  name: 'JWEDecryptionFailed'
```

**Reply by agusterodin** [reply]
Enabling the 'debug' option doesn't provide me with anything to assist in figuring out what the issue is.

I don't see anything get logged directly before the decryption error happens.

**Reply by agusterodin** [reply]
I also tried digging through the next-auth codebase and it is all way over my head.

**Reply by balazsorban44** [reply]
Could you start by adding a secret?

See https://github.com/nextauthjs/docs/pull/103/files#diff-05a54dcf3b29ce24cb3cb2c5c6780a680be7ffb8f73633bb329ba0f00d846696R78

(this will soon be merged, sorry for linking to a WIP PR)

**Reply by agusterodin** [reply]
Oh wow! That fixed it.

Is there any specific way you would like me to call out the fact that a secret is mandatory for this particular example to work?

I'm guessing that the secret will be ideally stored in .env. I can add it to the .env.example file and put a comment above that line with a link to the section of the docs in your PR.

Will the user get warned of missing secret in development environment console (when developing locally)?

Either way, I will continue to work on the updated example. Shouldn't be too difficult now that am aware that secret is required. I will strive to have a PR ready by the end of the weekend.

**Reply by balazsorban44** [reply]
This will soon be mandatory in production and show a warning in development:

https://github.com/nextauthjs/next-auth/pull/3143

This is a way to make apps more secure as well, so it was long overdue, to be honest...

**Reply by agusterodin** [reply]
I have been working on getting the remaining portion of the refresh token example working with v4. I noticed that the user profile claims (name, picture, etc) are present in the token that next-auth retrieves but not in the token refreshAccessToken gets.

refreshAccessToken in the example works by using fetch:
```
const url =
  'https://oauth2.googleapis.com/token?' +
  new URLSearchParams({
    client_id: process.env.GOOGLE_CLIENT_ID,
    client_secret: process.env.GOOGLE_CLIENT_SECRET,
    grant_type: 'refresh_token',
    refresh_token: token.refreshToken
  })

const response = await fetch(url, {
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded'
  },
  method: 'POST'
})

const refreshedTokens = await response.json()
```

I remember in another discussion (#3014) you mentioned that v4 is able to get a token that has profile claims whereas v3 can't because you transitioned over to using Open ID and powering the providers off of well-known.

My first instinct is to try to rewrite refreshAccessToken using the same openid-client that next-auth uses internally. Is this an advisable direction to go in?

**Reply by balazsorban44** [reply]
`refresh_token`s are usually used to get an updated `access_token` and in most cases (I believe not in Google's) also a new `refresh_token`.

To retrieve updated user claims, a provider might expose a `userinfo` endpoint, which you can think of as an API endpoint specific for user data. You usually send a valid `access_token` with the request, and you get back the user's claims as a JSON object. (Google has such an endpoint, check https://accounts.google.com/.well-known/openid-configuration)

**UPDATE:** Upon doing some research, you might get an `id_token` back as well:
 https://auth0.com/learn/refresh-tokens/

You will have to check the Google API docs. Currently, rotation is not supported built-in. If we do it, it will be done automatically, until then I cannot look up now how to do it for Google, sorry.

Since refreshing tokens right now is doing manual fetch calls, you might be better off with https://github.com/panva/node-openid-client, the package that `next-auth` uses under the hood. It actually has a refresh method for this purpose: https://github.com/panva/node-openid-client/blob/main/docs/README.md#clientrefreshrefreshtoken-extras

**Reply by agusterodin** [reply]
Cool, yeah next-auth v4 itself has no problem getting a JWT that contains all of the claims without requiring a separate request to a `userinfo` endpoint. That tells me that it should definitely be possible to somehow get a JWT like that inside my `refreshAccessToken` helper as well.

Sounds like using the `node-openid-client` (the same one that `next-auth` uses internally) inside my helper is the way to go!

**Reply by balazsorban44** [reply]
For now, yes. I hope we can add built-in support for token rotation soon. #3101 was necessary to be able to support it with confidence for both JWT and DB persisted sessions. Increasing security by default encryption was also a super important thing. v4 will become the base for many new features to come!

**Reply by agusterodin** [reply]
I rewrote the `refreshAccessToken` function to use `node-openid-client` like this:

```
  const googleIssuer = await Issuer.discover('https://accounts.google.com')
  const client = new googleIssuer.Client({
    client_id: process.env.GOOGLE_CLIENT_ID!,
    client_secret: process.env.GOOGLE_CLIENT_SECRET!,
    redirect_uris: [process.env.NEXTAUTH_URL!],
    response_types: ['id_token']
  })
  const tokenSet = await client.refresh(token.refreshToken || '')

  console.log('tokenSet', tokenSet)
```

I am able to successfully get an updated token set. Unfortunately the id token still doesn't contain profile claims.

Is there a way to see the exact request that next-auth performs out to the provider? I would like to compare requests to see if mine is missing anything that `next-auth`'s requests have that prevents Google from giving me an id token with profile claims.

**Reply by agusterodin** [reply]
Also, did you make any breakthroughs @mrkirchner and @ConciergeMento? If so, sharing would be really helpful so I can update the docs example.

**Reply by balazsorban44** [reply]
I don't think we log the outgoing requests at the moment. I'll look into if we can do that https://github.com/panva/node-openid-client/tree/main/docs#customizing-individual-http-requests

**Reply by agusterodin** [reply]
Unfortunate that I can't get profile claims when I send a request out myself using that library. 

Ended up change my backend API not to rely so much on profile claims (now it will only sync Google picture and display name periodically when those claims are present which is good enough for my purposes.

This is no blocker for updating the example repo. I should have that cleaned up and posted here in a zip file by the end of the day along with a PR for the docs. I prefer not to host the example repo from my GitHub account.

Worst case scenario if you don't want to host it from this account is that the zip file will be here for people to use and obviously the docs will be up to date!

Looking forward to the possibility of this library supporting refreshing out of the box. Also really hoping that the profile claims show up in those refreshed tokens 😀. Glad to be sponsoring this project. I really believe in what you are doing. You are making A LOT of people's lives easier.

**Reply by balazsorban44** [reply]
Thank you, that means a lot! What is and isn't returned by Google is kind of out of our control though, unfortunately. :slightly_frowning_face: 

You could check this endpoint by attaching an auth header and see if it returns what you are after. https://openidconnect.googleapis.com/v1/userinfo

`openid-client` also has a helper method for this: https://github.com/panva/node-openid-client/blob/main/docs/README.md#clientuserinfoaccesstoken-options

**Reply by agusterodin** [reply]
Put a PR in for the docs fix: https://github.com/nextauthjs/docs/pull/117. Attached a working v4 example to it if you decide you want to use it as well. This combined with your soon to come "secret required" warning fixes should address everything.

**Reply by ndom91** [reply]
Thanks a lot @agusterodin, we've uploaded this at [nextauthjs/next-auth-refresh-token-example](https://github.com/nextauthjs/next-auth-refresh-token-example) and made sure to thank you in the README and list you as author in the `package.json` :heart: :tada:

**Comment by mrkirchner** [comment]
@balazsorban44 one thing i noticed is if i use authorizationUrl instead of authorization i tend to not get JWEDecryptionFailed. Ive tried it in this app and another which doesnt really need the google specific access_type or connection_scope stuff.

```
   DOES NOT WORK - I never get refresh_token but i also dont get any JWEDecryptionFailed messages either.

    Auth0Provider({
      clientId: process.env.AUTH0_CLIENT_ID as string,
      clientSecret: process.env.AUTH0_CLIENT_SECRET as string,
      issuer: process.env.AUTH0_ISSUER,
      authorizationUrl: `${process.env.AUTH0_ISSUER}/authorize?response_type=code&prompt=consent&connection_scope=https://www.googleapis.com/auth/calendar&access_type=offline`,
    }),


vs 

   WORKS KIND OF - in the sense that i get Refresh Token on Identity but i get JWEDecryptionFailed: decryption operation failed any time i get session

 Auth0Provider({
      clientId: process.env.AUTH0_CLIENT_ID as string,
      clientSecret: process.env.AUTH0_CLIENT_SECRET as string,
      issuer: process.env.AUTH0_ISSUER,
      authorization: `${process.env.AUTH0_ISSUER}/authorize?response_type=code&prompt=consent&connection_scope=https://www.googleapis.com/auth/calendar&access_type=offline`,
    }),

vs 

   WORKS KIND OF - in the sense that i get Refresh Token on Identity but i get JWEDecryptionFailed: decryption operation failed any time i get session

    Auth0Provider({
      clientId: process.env.AUTH0_CLIENT_ID as string,
      clientSecret: process.env.AUTH0_CLIENT_SECRET as string,
      issuer: process.env.AUTH0_ISSUER,
      authorization: {
        params: {
          prompt: "consent",
          response_type: "code",
          access_type: "offline",
          connection_scope:
            "https://www.googleapis.com/auth/calendar",
        },
    }}),
```
Any insight on what is going on between these 3?  Thanks for your help.

**Comment by balazsorban44** [comment]
authorizationUrl does nothing in v4. there is no reference to it in the core
